cloudsdk/src/main/java/io/particle/android/sdk/cloud/ParticleDevice.java

```diff
@@ -471,20 +479,105 @@ private void resetFlashingState() {
     private void performFlashingChange(FlashingChange flashingChange) throws ParticleCloudException {
         try {
             flashingChange.executeFlashingChange();
-            isFlashing = true;
-            cloud.notifyDeviceChanged();
-            // Gross.  We're using this "run delayed" hack just for the *scheduling* aspect
-            // of this, and then we're just telling the scheduled runnable to drop right back to a
-            // background thread so we don't call resetFlashingState() on the main thread.
-            // Still, I don't want to introduce a whole scheduled executor setup *just for this*,
-            // or write something that just sits in a Thread.sleep(), hogging a whole thread when
-            // more important work could be getting blocked.
-            EZ.runOnMainThreadDelayed(30000, () -> EZ.runAsync(() -> resetFlashingState()));
-        } catch (RetrofitError e) {
+            //listens for flashing event, on success unsubscribe from listening.
+            subscribeToSystemEvent("spark/flash/status", new SimpleParticleEventHandler() {
+                @Override
+                public void onEvent(String eventName, ParticleEvent particleEvent) {
+                    if (particleEvent.dataPayload.equals("success")) {
+                        isFlashing = false;
+                        try {
+                            ParticleDevice.this.refresh();
+                            unsubscribeFromEvents(this);
+                        } catch (ParticleCloudException e) {
+                            // not much else we can really do here...
+                            log.w("Unable to reset flashing state for %s" + deviceState.deviceId, e);
+                        }
+                    } else {
+                        isFlashing = true;
+                    }
+                    cloud.notifyDeviceChanged();
+                }
+            });
+        } catch (RetrofitError | IOException e) {
             throw new ParticleCloudException(e);
         }
     }
 
+    /**
+     * Subscribes to system events of current device. Events emitted to EventBus listener.
+     *
+     * @throws ParticleCloudException Failure to subscribe to system events.
+     * @see <a href="https://github.com/greenrobot/EventBus">EventBus</a>
+     */
+    public void subscribeToSystemEvents() throws ParticleCloudException {
```

Comment:
Since this is mutating a collection, you should at least annotate this `@MainThread` (and consider using `CopyOnWriteArrayList` anyhow)